c++/src/kj/async-io-unix.c++

```diff
@@ -228,6 +228,149 @@ public:
     return promise.attach(kj::mv(fds), kj::mv(streams));
   }
 
+#if __linux__
+  Maybe<Promise<uint64_t>> tryPumpFrom(
+      AsyncInputStream& inputParam, uint64_t amount = kj::maxValue) override {
+    AsyncStreamFd& input = KJ_UNWRAP_OR_RETURN(
+        kj::dynamicDowncastIfAvailable<AsyncStreamFd>(inputParam), nullptr);
+
+    // The input is another AsyncStreamFd, so perhaps we can do an optimized pump with splice().
+
+    // Before we resort to a bunch of syscalls, let's try to see if the pump is small and able to
+    // be fully satisfied immediately. This optimizes for the case of small streams, e.g. a short
+    // HTTP body.
+
+    byte buffer[4096];
+    size_t pos = 0;
+    size_t initialAmount = kj::min(sizeof(buffer), amount);
+
+    bool eof = false;
+
+    // Read into the buffer until it's full or there are no bytes available. Note that we'd expect
+    // one call to read() will pull as much data out of the socket as possible (up to our buffer
+    // size), so you might think the loop is unnecessary. The reason we want to do a second read(),
+    // though, is to find out if we're at EOF or merely waiting for more data. In the EOF case,
+    // we can end the pump early without splicing.
+    while (pos < initialAmount) {
+      ssize_t n;
+      KJ_NONBLOCKING_SYSCALL(n = ::read(input.fd, buffer + pos, initialAmount - pos));
+      if (n <= 0) {
+        eof = n == 0;
+        break;
+      }
+      pos += n;
+    }
+
+    // Write the bytes that we just read back out to the output.
+    {
+      ssize_t n;
+      KJ_NONBLOCKING_SYSCALL(n = ::write(fd, buffer, pos));
+      if (n < 0) n = 0;  // treat EAGAIN as "zero bytes written"
+      if (size_t(n) < pos) {
+        // Oh crap, the output buffer is full. This should be rare. But, now we're going to have
+        // to copy the remaining bytes into the heap to do an async write.
+        auto leftover = kj::heapArray<byte>(buffer + n, pos - n);
+        auto promise = write(leftover.begin(), leftover.size());
+        promise = promise.attach(kj::mv(leftover));
+        if (eof || pos == initialAmount) {
```

Comment:
Hmmmmm... it seems like I meant to write `pos == amount`, i.e. check whether we fulfilled the entire pump.

But surely this bug should have been caught by tests... let me investigate.

Reply:
Yeah good catch... tests missed it because they initiated the pump too early. Fixed in: 356f7d647afe9c7323dc30ecfbcd32f91ed27812

Weird that I didn't see higher-level Workers tests hit this, I had expected this to be a common case.

Reply:
Same, I thought for sure I was missing some subtlety here.